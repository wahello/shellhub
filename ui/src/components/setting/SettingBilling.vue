<template>
  <v-container>
    <v-row
      align="center"
      justify="center"
      class="mt-4"
    >
      <v-col
        sm="8"
      >
        <SettingOwnerInfo
          :is-owner="isOwner"
          data-test="settingOwnerInfo-component"
        />

        <div
          v-if="isOwner"
          data-test="content-div"
        >
          <v-row>
            <v-col>
              <h3>
                Subscription info
              </h3>
            </v-col>

            <v-spacer />

            <v-col
              v-if="isOwner && !active"
              md="auto"
              class="ml-auto"
            >
              <PaymentMethod
                type-operation="subscription"
                data-test="subscriptionPaymentMethod-component"
                @update="getSubscriptionInfo()"
              />
            </v-col>
          </v-row>

          <div class="mt-6 pl-4 pr-4">
            <div v-if="isOwner && !active">
              <p>
                Plan: <b> Free </b>
              </p>

              <p>
                Description: you can only add 3 devices.
              </p>
            </div>

            <div v-else-if="isOwner && active">
              <p>
                Plan: <b> Premium usage </b>
              </p>

              <p>
                Description: In this plan, the amount is charged according to the number of
                devices used.
              </p>
            </div>
          </div>
        </div>

        <div
          v-if="isOwner && active && renderData"
          class="mt-4 mb-4"
          data-test="subscriptionActive-div"
        >
          <v-divider />
          <v-divider />

          <div class="mt-6">
            <v-row>
              <v-col>
                <h3>
                  Next bill
                </h3>
              </v-col>
            </v-row>

            <div class="mt-6 pl-4 pr-4">
              <p>
                Date: <b> {{ infoBillingData.periodEnd | formatDateWithoutDayAndHours }} </b>
              </p>

              <p>
                Estimated total:
                <b v-if="renderData">
                  {{ infoBillingData.nextPaymentDue | formatCurrency }}
                </b>
              </p>
            </div>
          </div>

          <v-divider />
          <v-divider />

          <div class="mt-6">
            <v-row>
              <v-col>
                <h3>
                  Payment method details
                </h3>
              </v-col>

              <v-spacer />

              <v-col
                md="auto"
                class="ml-auto"
              >
                <PaymentMethod
                  type-operation="update"
                  data-test="updatePaymentMethod-component"
                  @update="getSubscriptionInfo()"
                />
              </v-col>
            </v-row>

            <div class="mt-5 pl-4 pr-4">
              <p>
                <BillingIcon
                  v-if="renderData"
                  :icon-name="cardBillingData.brand"
                  data-test="billingIcon-component"
                />
                {{ cardBillingData.expMonth }}/{{ cardBillingData.expYear }} -
                {{ cardBillingData.last4 }}
              </p>
            </div>
          </div>

          <v-divider />
          <v-divider />

          <div class="mt-6">
            <v-row>
              <v-col>
                <h3>
                  Cancel Subscription
                </h3>
              </v-col>
            </v-row>

            <div class="mt-2 pl-4">
              <v-row>
                <v-col>
                  <p>
                    When canceling subscription, you may lose access to devices.
                  </p>
                </v-col>

                <v-spacer />

                <v-col
                  md="auto"
                  class="ml-auto"
                >
                  <BillingCancel
                    v-if="renderData"
                    :next-payment-due="infoBillingData.nextPaymentDue"
                    @update="updateNamespace()"
                  />
                </v-col>
              </v-row>
            </div>
          </div>
        </div>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>

import SettingOwnerInfo from '@/components/setting/SettingOwnerInfo';
import BillingCancel from '@/components/billing/BillingCancel';
import PaymentMethod from '@/components/billing/BillingDialogPaymentMethod';
import BillingIcon from '@/components/billing/BillingIcon';

import { formatDateWithoutDayAndHours } from '@/components/filter/date';
import formatCurrency from '@/components/filter/currency';

export default {
  name: 'SettingBilling',

  components: {
    SettingOwnerInfo,
    BillingCancel,
    PaymentMethod,
    BillingIcon,
  },

  filters: { formatDateWithoutDayAndHours, formatCurrency },

  data() {
    return {
      card: null,
      elements: null,
      billingData: { info: Object, card: Object },
      renderData: false,
    };
  },

  computed: {
    active() {
      return this.$store.getters['billing/active'];
    },

    billing() {
      return this.$store.getters['billing/get'];
    },

    isOwner() {
      return this.$store.getters['namespaces/owner'];
    },

    infoBillingData() {
      return this.billingData.info;
    },

    cardBillingData() {
      return this.billingData.card;
    },
  },

  watch: {
    isOwner(status) {
      if (status) {
        this.stripeData();
      }
    },
  },

  created() {
    this.updateNamespace();
  },

  mounted() {
    if (this.isOwner) {
      this.stripeData();
    }
  },

  methods: {
    stripeData() {
      this.mountStripeElements();

      if (this.active) {
        this.getSubscriptionInfo();
      }
    },

    mountStripeElements() {
      this.elements = this.$stripe.elements();
      this.card = this.elements.create('card');
    },

    formatSubscriptionData(data) {
      const latestInvoice = data.latest_invoice;
      const upcomingInvoice = data.upcoming_invoice;
      const productDescription = data.product_description;
      const { card } = data;

      this.billingData.info = {
        periodEnd: this.billing.current_period_end,
        description: productDescription,
        latestPaymentDue: latestInvoice.amount_due,
        latestPaymentPaid: latestInvoice.amount_paid,
        nextPaymentDue: upcomingInvoice.amount_due,
        nextPaymentPaid: upcomingInvoice.amount_paid,
      };

      this.billingData.card = {
        brand: card.brand,
        expYear: card.exp_year,
        expMonth: card.exp_month,
        last4: card.last4,
      };
    },

    async getSubscriptionInfo() {
      if (this.active) {
        try {
          const data = await this.$store.dispatch('billing/getSubscription');
          await this.formatSubscriptionData(data);

          this.renderData = true;
        } catch {
          this.renderData = false;
          this.$store.dispatch('snackbar/showSnackbarErrorDefault');
        }
      }
    },

    async updateNamespace() {
      try {
        await this.$store.dispatch('namespaces/get', localStorage.getItem('tenant'));
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorLoading', this.$errors.snackbar.namespaceLoad);
      }
    },
  },
};

</script>
