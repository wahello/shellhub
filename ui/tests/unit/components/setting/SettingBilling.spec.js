import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import flushPromises from 'flush-promises';
import SettingBilling from '@/components/setting/SettingBilling';

describe('SettingBilling', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);

  let wrapper;
  const owner = true;

  const inactiveBilling = {
    active: false,
    current_period_end: 0,
    customer_id: '',
    subscription_id: '',
    payment_method_id: '',
  };

  const activeBilling = {
    active: true,
    current_period_end: 12121,
    customer_id: 'cus_123',
    subscription_id: 'subs_123',
    payment_method_id: 'pm_123',
  };

  const stripeData = {
    latest_invoice: { amount_due: 0, amount_paid: 0 },
    upcoming_invoice: { amount_due: 0, amount_paid: 0 },
    product_description: 'Premium usage',
    card: {
      brand: 'visa', exp_year: 2024, exp_month: 4, last4: '4242',
    },
  };

  const billingData = {
    info: {
      description: stripeData.product_description,
      latestPaymentDue: stripeData.latest_invoice.amount_due,
      latestPaymentPaid: stripeData.latest_invoice.amount_paid,
      nextPaymentDue: stripeData.upcoming_invoice.amount_due,
      nextPaymentPaid: stripeData.upcoming_invoice.amount_paid,
    },
    card: stripeData.card,
  };

  const storeOwnerWithoutSubscription = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling: inactiveBilling.active,
      billing: inactiveBilling,
      owner,
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'billing/get': (state) => state.billing,
      'namespaces/owner': (state) => state.owner,
    },
    actions: {
      'billing/getSubscription': () => stripeData,
      'namespaces/get': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
      'snackbar/showSnackbarErrorDefault': () => {},
    },
  });

  const storeOwnerWithSubscription = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling: activeBilling.active,
      billing: activeBilling,
      owner,
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'billing/get': (state) => state.billing,
      'namespaces/owner': (state) => state.owner,
    },
    actions: {
      'billing/getSubscription': () => stripeData,
      'billing/cancelSubscription': () => {},
      'namespaces/get': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
      'snackbar/showSnackbarErrorDefault': () => {},
    },
  });

  const storeNotOwner = new Vuex.Store({
    namespaced: true,
    state: {
      stateBilling: activeBilling.active,
      billing: activeBilling,
      owner: !owner,
    },
    getters: {
      'billing/active': (state) => state.stateBilling,
      'billing/get': (state) => state.billing,
      'namespaces/owner': (state) => state.owner,
    },
    actions: {
      'billing/getSubscription': () => {},
      'billing/cancelSubscription': () => {},
      'namespaces/get': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorAction': () => {},
      'snackbar/showSnackbarErrorDefault': () => {},
    },
  });

  ///////
  // In this case, when the user owns the namespace and the focus of
  // the test is the unrealized owner of the subscription.
  ///////

  describe('Namespace owner without subscription', () => {
    beforeEach(() => {
      wrapper = shallowMount(SettingBilling, {
        store: storeOwnerWithoutSubscription,
        localVue,
        stubs: ['fragment'],
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.billingData).toEqual({ card: Object, info: Object });
      expect(wrapper.vm.renderData).toEqual(false);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.active).toEqual(inactiveBilling.active);
      expect(wrapper.vm.billing).toEqual(inactiveBilling);
      expect(wrapper.vm.isOwner).toEqual(owner);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', () => {
      expect(wrapper.find('[data-test="settingOwnerInfo-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="subscriptionPaymentMethod-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="updatePaymentMethod-component"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="billingIcon-component"]').exists()).toBe(false);
    });
    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="content-div"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="subscriptionActive-div"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(false);
    });
  });

  ///////
  // In this case, when the user owns the namespace and the focus of
  // the test is subscription is already done.
  ///////

  describe('Namespace owner with subscription', () => {
    beforeEach(async () => {
      wrapper = shallowMount(SettingBilling, {
        store: storeOwnerWithSubscription,
        localVue,
        stubs: ['fragment'],
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });

      wrapper.setData({
        renderData: true,
        billingData,
      });
      await flushPromises();
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.renderData).toEqual(true);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.active).toEqual(activeBilling.active);
      expect(wrapper.vm.billing).toEqual(activeBilling);
      expect(wrapper.vm.isOwner).toEqual(owner);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', () => {
      expect(wrapper.find('[data-test="settingOwnerInfo-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="subscriptionPaymentMethod-component"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="updatePaymentMethod-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="billingIcon-component"]').exists()).toBe(true);
    });
    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="content-div"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="subscriptionActive-div"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(false);
    });
  });

  // ///////
  // // In this case, when the user doesn't owns the namespace.
  // ///////

  describe('Doesn\'t own the namespace', () => {
    beforeEach(async () => {
      wrapper = shallowMount(SettingBilling, {
        store: storeNotOwner,
        localVue,
        stubs: ['fragment'],
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });

      wrapper.setData({ renderData: true });
      await flushPromises();
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Compare data with default value', () => {
      expect(wrapper.vm.billingData).toEqual({ card: Object, info: Object });
      expect(wrapper.vm.renderData).toEqual(true);
    });
    it('Process data in the computed', () => {
      expect(wrapper.vm.active).toEqual(activeBilling.active);
      expect(wrapper.vm.billing).toEqual(activeBilling);
      expect(wrapper.vm.isOwner).toEqual(!owner);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with components', () => {
      expect(wrapper.find('[data-test="settingOwnerInfo-component"]').exists()).toBe(true);
      expect(wrapper.find('[data-test="subscriptionPaymentMethod-component"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="updatePaymentMethod-component"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="billingIcon-component"]').exists()).toBe(false);
    });
    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="content-div"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="subscriptionActive-div"]').exists()).toBe(false);
      expect(wrapper.find('[data-test="cancel-btn"]').exists()).toBe(false);
    });
  });
});
