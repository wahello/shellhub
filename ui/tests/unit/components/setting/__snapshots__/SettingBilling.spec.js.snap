// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingBilling Doesn't own the namespace Renders the component 1`] = `
<v-container-stub tag="div">
  <v-row-stub tag="div" align="center" justify="center" class="mt-4">
    <v-col-stub sm="8" tag="div">
      <settingownerinfo-stub data-test="settingOwnerInfo-component"></settingownerinfo-stub>
      <!---->
      <!---->
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;

exports[`SettingBilling Namespace owner with subscription Renders the component 1`] = `
<v-container-stub tag="div">
  <v-row-stub tag="div" align="center" justify="center" class="mt-4">
    <v-col-stub sm="8" tag="div">
      <settingownerinfo-stub isowner="true" data-test="settingOwnerInfo-component"></settingownerinfo-stub>
      <div data-test="content-div">
        <v-row-stub tag="div">
          <v-col-stub tag="div">
            <h3>
              Subscription info
            </h3>
          </v-col-stub>
          <v-spacer-stub></v-spacer-stub>
          <!---->
        </v-row-stub>
        <div class="mt-6 pl-4 pr-4">
          <div>
            <p>
              Plan: <b> Premium usage </b></p>
            <p>
              Description: In this plan, the amount is charged according to the number of
              devices used.
            </p>
          </div>
        </div>
      </div>
      <div data-test="subscriptionActive-div" class="mt-4 mb-4">
        <v-divider-stub></v-divider-stub>
        <v-divider-stub></v-divider-stub>
        <div class="mt-6">
          <v-row-stub tag="div">
            <v-col-stub tag="div">
              <h3>
                Next bill
              </h3>
            </v-col-stub>
          </v-row-stub>
          <div class="mt-6 pl-4 pr-4">
            <p>
              Date: <b> January 1st 1970 </b></p>
            <p>
              Estimated total:
              <b>
                $0.00
              </b>
            </p>
          </div>
        </div>
        <v-divider-stub></v-divider-stub>
        <v-divider-stub></v-divider-stub>
        <div class="mt-6">
          <v-row-stub tag="div">
            <v-col-stub tag="div">
              <h3>
                Payment method details
              </h3>
            </v-col-stub>
            <v-spacer-stub></v-spacer-stub>
            <v-col-stub md="auto" tag="div" class="ml-auto">
              <paymentmethod-stub typeoperation="update" data-test="updatePaymentMethod-component"></paymentmethod-stub>
            </v-col-stub>
          </v-row-stub>
          <div class="mt-5 pl-4 pr-4">
            <p>
              <billingicon-stub iconname="visa" data-test="billingIcon-component"></billingicon-stub>
              4/2024 -
              4242
            </p>
          </div>
        </div>
        <v-divider-stub></v-divider-stub>
        <v-divider-stub></v-divider-stub>
        <div class="mt-6">
          <v-row-stub tag="div">
            <v-col-stub tag="div">
              <h3>
                Cancel Subscription
              </h3>
            </v-col-stub>
          </v-row-stub>
          <div class="mt-2 pl-4">
            <v-row-stub tag="div">
              <v-col-stub tag="div">
                <p>
                  When canceling subscription, you may lose access to devices.
                </p>
              </v-col-stub>
              <v-spacer-stub></v-spacer-stub>
              <v-col-stub md="auto" tag="div" class="ml-auto">
                <billingcancel-stub nextpaymentdue="0"></billingcancel-stub>
              </v-col-stub>
            </v-row-stub>
          </div>
        </div>
      </div>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;

exports[`SettingBilling Namespace owner without subscription Renders the component 1`] = `
<v-container-stub tag="div">
  <v-row-stub tag="div" align="center" justify="center" class="mt-4">
    <v-col-stub sm="8" tag="div">
      <settingownerinfo-stub isowner="true" data-test="settingOwnerInfo-component"></settingownerinfo-stub>
      <div data-test="content-div">
        <v-row-stub tag="div">
          <v-col-stub tag="div">
            <h3>
              Subscription info
            </h3>
          </v-col-stub>
          <v-spacer-stub></v-spacer-stub>
          <v-col-stub md="auto" tag="div" class="ml-auto">
            <paymentmethod-stub typeoperation="subscription" data-test="subscriptionPaymentMethod-component"></paymentmethod-stub>
          </v-col-stub>
        </v-row-stub>
        <div class="mt-6 pl-4 pr-4">
          <div>
            <p>
              Plan: <b> Free </b></p>
            <p>
              Description: you can only add 3 devices.
            </p>
          </div>
        </div>
      </div>
      <!---->
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
